### PR TITLE
[gateway] http extauth sspi ntlm

### PR DESCRIPTION
**Pull request by akarl10**
This is an implementation of HTTP_EXTAUTH_SSPI_NTLM.

it can be enabled with `/gt:http|auto,extauth-sspi-ntlm`

### CODE REVIEW

File: libfreerdp/core/gateway/rdg.c [hunk 1]
```diff
@@ -1000,6 +1000,36 @@ static BOOL rdg_send_handshake(rdpRdg* rdg)
 	return status;
 }
 
+static BOOL rdg_send_extauth_sspi(rdpRdg* rdg)
+{
+	wStream* s;
+	BOOL status;
+	UINT32 packetSize = 8 + 4 + 2;
+
+	const SecBuffer* authToken = credssp_auth_get_output_buffer(rdg->auth);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1009]: `WINPR_ASSERT(rdg);` for new code (keep on the safe side)

File: libfreerdp/core/gateway/rdg.c [hunk 2]
```diff
@@ -1000,6 +1000,36 @@ static BOOL rdg_send_handshake(rdpRdg* rdg)
...
+	s = Stream_New(NULL, packetSize);
+
+	if (!s)
+		return FALSE;
+
+	Stream_Write_UINT16(s, PKT_TYPE_EXTENDED_AUTH_MSG); /* Type (2 bytes) */
+	Stream_Write_UINT16(s, 0);                          /* Reserved (2 bytes) */
+	Stream_Write_UINT32(s, packetSize);                 /* PacketLength (4 bytes) */
+	Stream_Write_UINT32(s, ERROR_SUCCESS);              /* Error code */
+	Stream_Write_UINT16(s, (unsigned short)authToken->cbBuffer);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1023]: `UINT16`

File: libfreerdp/core/gateway/rdg.c [hunk 3]
```diff
@@ -1000,6 +1000,36 @@ static BOOL rdg_send_handshake(rdpRdg* rdg)
...
+	Stream_Write(s, authToken->pvBuffer, (unsigned int)authToken->cbBuffer);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1024]: not required, the argument is `size_t` and `cbBuffer` is `ULONG` so it is always greater or equal

File: libfreerdp/core/gateway/rdg.c [hunk 4]
```diff
@@ -1443,6 +1476,50 @@ static BOOL rdg_process_tunnel_authorization_response(rdpRdg* rdg, wStream* s)
 	return rdg_send_channel_create(rdg);
 }
 
+static BOOL rdg_process_extauth_sspi(rdpRdg* rdg, wStream* s)
+{
+	UINT32 errorCode;
+	UINT16 authBlobLen;
+	SecBuffer authToken = { 0 };
+	BYTE* authTokenData = NULL;
+
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1485]: add `WINPR_ASSERT(rdg);`

File: libfreerdp/core/gateway/rdg.c [hunk 5]
```diff
@@ -1880,7 +1965,11 @@ static BOOL rdg_establish_data_connection(rdpRdg* rdg, rdpTls* tls, const char*
 			rdg->transferEncoding.isWebsocketTransport = TRUE;
 			rdg->transferEncoding.context.websocket.state = WebsocketStateOpcodeAndFin;
 			rdg->transferEncoding.context.websocket.responseStreamBuffer = NULL;
-
+			if (rdg->extAuth == HTTP_EXTENDED_AUTH_SSPI_NTLM)
+			{
+				if (!rdg_auth_init(rdg, tls, "NTLM"))
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1970]: Replace with `NTLM_SSP_NAME` (this version will fail on windows)

**Review by akallabeth — CHANGES_REQUESTED** [review]
Overall LGTM, some small changes required.
Nice contribution, thank you!

File: libfreerdp/core/gateway/rdg.c [hunk 6]
```diff
@@ -2592,7 +2695,7 @@ rdpRdg* rdg_new(rdpContext* context)
 		rdg->state = RDG_CLIENT_STATE_INITIAL;
 		rdg->context = context;
 		rdg->settings = rdg->context->settings;
-		rdg->extAuth = HTTP_EXTENDED_AUTH_NONE;
+		rdg->extAuth = (rdg->settings->GatewayHttpExtAuthSspiNtln?HTTP_EXTENDED_AUTH_SSPI_NTLM:HTTP_EXTENDED_AUTH_NONE);
```
Comment (hardening) [inline libfreerdp/core/gateway/rdg.c:2698]: code style: don't pack things this way (run `clang-format` on the file)

File: libfreerdp/core/gateway/rdg.c [hunk 7]
```diff
@@ -1443,6 +1478,52 @@ static BOOL rdg_process_tunnel_authorization_response(rdpRdg* rdg, wStream* s)
...
+		if (credssp_auth_is_complete(rdg->auth))
+		{
+			credssp_auth_free(rdg->auth);
+			rdg->auth = NULL;
+			return rdg_send_tunnel_request(rdg);
+		}
+		return FALSE;
+	}
+
+	authTokenData = malloc(sizeof(BYTE) * authBlobLen);
```
Comment (hardening) [inline libfreerdp/core/gateway/rdg.c:1510]: `sizeof(BYTE)` is useless

**Review by hardening — APPROVED** [review]
Except my remarks LGTM

**Comment by akarl10** [comment]
> Refer to this link for build results (access rights to CI server needed): https://ci.freerdp.com//job/PullRequestTester/8395/

sorry, did clang-format on the mentioned files...

File: libfreerdp/core/gateway/rdg.c [hunk 8]
```diff
@@ -1443,6 +1478,52 @@ static BOOL rdg_process_tunnel_authorization_response(rdpRdg* rdg, wStream* s)
...
+	BYTE* authTokenData = NULL;
+
+	WINPR_ASSERT(rdg);
+
+	Stream_Read_UINT32(s, errorCode);
+	Stream_Read_UINT16(s, authBlobLen);
+
+	if (errorCode != ERROR_SUCCESS)
+	{
+		WLog_ERR(TAG, "[%s] EXTAUTH_SSPI_NTLM error 0x%x", __FUNCTION__, errorCode);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.c:1495]: stringify error code with `GetSecurityStatusString`

**Review by akallabeth — APPROVED** [review]
Except for the debug log remark LGTM

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Pull request merged**